Documentation/bpf.rst

```diff
@@ -989,6 +1079,85 @@ currently unsupported, too.
 Furthermore, compilation from BPF assembly (e.g. ``llvm-mc xdp-example.S -arch bpf -filetype=obj -o xdp-example.o``)
 is currently not supported either due to missing BPF assembly parser.
 
+LLVM by default uses the BPF base instruction set for generating code
+in order to make sure that the generated object file can also be loaded
+with older kernels such as long-term stable kernels (e.g. 4.9+).
+
+However, LLVM has a ``-mcpu`` selector for the BPF back end in order to
+select different versions of the BPF instruction set, namely instruction
+set extensions on top of the BPF base instruction set in order to generate
+more efficient and smaller code.
+
+Available ``-mcpu`` options can be queried through:
+
+::
+
+    $ llc -march bpf -mcpu=help
+    Available CPUs for this target:
+
+      generic - Select the generic processor.
+      probe   - Select the probe processor.
+      v1      - Select the v1 processor.
+      v2      - Select the v2 processor.
+    [...]
+
+The ``generic`` processor is the default processor, which is also the
+base instruction set ``v1`` of BPF. Options ``v1`` and ``v2`` are typically
+useful in an environment where the BPF program is being cross compiled
+and the target host where the program is loaded differs from the one
+where it is compiled (and thus available BPF kernel features might differ
+as well).
+
+The recommended ``-mcpu`` option which is also used by Cilium internally is
+``-mcpu=probe``! Here, the LLVM BPF back end queries the kernel for availability
+of BPF instruction set extensions and when found available, LLVM will use
+them for compiling the BPF program whenever appropriate.
+
+A full command line example with llc's ``-mcpu=probe``:
+
+::
+
+    $ clang -O2 -Wall -target bpf -emit-llvm -c xdp-example.c -o xdp-example.bc
+    $ llc xdp-example.bc -march=bpf -mcpu=probe -filetype=obj -o xdp-example.o
+
+Generally, LLVM IR generation is mostly architecture independent. There
```

Comment:
"generally" and "mostly" here serve much the same purpose, so we probably only need one of them.